File: coding/problems_asked_in_my_interviews/longest_common_prefix.cpp
// Leetcode problem link - https://leetcode.com/problems/longest-common-prefix

// Solutions -

// Approach 1
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        string baseString = strs[0], finalString = "";
        for (int i = 0; i < strs[0].size(); i++) {
            char baseChar = strs[0][i];
            for (int j = 1; j < strs.size(); j++) {
                if (i == strs[j].size() || strs[j][i] != baseChar) {
                    return finalString;
                }
            }
            finalString += baseChar;
        }
        return finalString;
    }
};
// Time complexity - O(m*n) [assuming size of the base string as 'm', and size of array as 'n']
// Space complexity - O(m*n)
// Auxiliary space - O(1)

File: coding/neetcode/longest_common_prefix.py
# Problem - https://leetcode.com/problems/longest-common-prefix/

# Solutions

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        base_string = strs[0]
        common_prefix = ''

        for i in range(len(base_string)):
            for j in range(1, len(strs)):
                if i == len(strs[j]) or strs[j][i] != base_string[i]:
                    return common_prefix

            common_prefix += base_string[i]
        
        return common_prefix
# Time complexity - O(m*(n-1)) - O(m*n) where m = size of base string and n = size of list of strings
# Space complexity - O(n)
# Auxiliary space - O(1)
